### PR TITLE
SpectrumInfo rollout: algorithms SetInstrumentParameter to PredictPeaks

### LINKED ISSUE #18554: SpectrumInfo and DetectorInfo rollout: algorithms <SetInstrumentParameter to PredictPeaks>
This issue is part of the SpectrumInfo and DetectorInfo rollout, see https://github.com/mantidproject/mantid/issues/17743.
Algorithms covered in this issue are:
[SetInstrumentParameter](http://docs.mantidproject.org/nightly/algorithms/SetInstrumentParameter-v1.html)
[RingProfile](http://docs.mantidproject.org/nightly/algorithms/RingProfile-v1.html)
[AddPeakHKL](http://docs.mantidproject.org/nightly/algorithms/AddPeakHKL-v1.html)
[PeaksInRegion](http://docs.mantidproject.org/nightly/algorithms/PeaksInRegion-v1.html)
[IntegratePeakTimeSlices](http://docs.mantidproject.org/nightly/algorithms/IntegratePeakTimeSlices-v1.html)
[FindSXPeaks](http://docs.mantidproject.org/nightly/algorithms/FindSXPeaks-v1.html)
[PredictPeaks](http://docs.mantidproject.org/nightly/algorithms/PredictPeaks-v1.html)

### PR DESCRIPTION
**Pull request by soininen**
## Description of work.
This PR is part of refactoring effort to use the new interface of the SpectrumInfo and DetectorInfo modules.
See https://github.com/mantidproject/mantid/issues/17743 for details.

The following algorithms were refactored to use SpectrumInfo/DetectorInfo:
[SetInstrumentParameter](http://docs.mantidproject.org/nightly/algorithms/SetInstrumentParameter-v1.html) (no changes)
[RingProfile](http://docs.mantidproject.org/nightly/algorithms/RingProfile-v1.html) (no changes)
[AddPeakHKL](http://docs.mantidproject.org/nightly/algorithms/AddPeakHKL-v1.html) (no changes)
[PeaksInRegion](http://docs.mantidproject.org/nightly/algorithms/PeaksInRegion-v1.html)
[IntegratePeakTimeSlices](http://docs.mantidproject.org/nightly/algorithms/IntegratePeakTimeSlices-v1.html)
[FindSXPeaks](http://docs.mantidproject.org/nightly/algorithms/FindSXPeaks-v1.html)
[PredictPeaks](http://docs.mantidproject.org/nightly/algorithms/PredictPeaks-v1.html) (no changes)

## To test.

Code review.

Fixes #18554.

Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Crystal/test/PeaksInRegionTest.h [hunk 1]
```diff
@@ -30,7 +31,7 @@ class PeaksInRegionTest : public CxxTest::TestSuite {
                        double yMaxFromPeak, double zMinFromPeak,
                        double zMaxFromPeak) {
     PeaksWorkspace_sptr ws = WorkspaceCreationHelper::createPeaksWorkspace(1);
-    auto detectorIds = ws->getInstrument()->getDetectorIDs();
+    const auto detectorIds = ws->detectorInfo().detectorIDs();
```
Comment (SimonHeybrock) [inline Framework/Crystal/test/PeaksInRegionTest.h]: reference

File: Framework/Crystal/inc/MantidCrystal/FindSXPeaks.h [hunk 2]
```diff
@@ -39,12 +37,29 @@ struct DLLExport SXPeak {
       throw std::invalid_argument(
           "SXPeak: Cannot have zero sized spectral list");
     }
-    if (Ltot < 0) {
+    if (!spectrumInfo.hasUniqueDetector(_wsIndex)) {
+      throw std::invalid_argument("SXPeak: Spectrum at ws index " +
+                                  std::to_string(wsIndex) +
+                                  " has more than a single detector");
```
Comment (SimonHeybrock) [inline Framework/Crystal/inc/MantidCrystal/FindSXPeaks.h]: I can't find the equivalent of this check/exception in the original version?

File: Framework/Crystal/src/IntegratePeakTimeSlices.cpp [hunk 3]
```diff
@@ -259,18 +259,14 @@ void IntegratePeakTimeSlices::exec() {
...
   TableWorkspace_sptr TabWS = boost::make_shared<TableWorkspace>(0);
 
   //----------------------------- get Peak extents
   //------------------------------
   try {
-
-    // Find the workspace index for this detector ID
-    detid2index_map::const_iterator it = m_wi_to_detid_map.find(detID);
-    size_t wsIndx = (it->second);
+    const size_t wsIndx = detectorInfo.indexOf(detID);
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/IntegratePeakTimeSlices.cpp]: This is not what you need. The index returned by `indexOf` is an index for the detector, but you need an index for the detector group (i.e., the index of spectrum which contains the detector). I think you need to keep the old way of obtaining the index.

File: Framework/Crystal/src/IntegratePeakTimeSlices.cpp [hunk 4]
```diff
@@ -1570,13 +1566,14 @@ void IntegratePeakTimeSlices::SetUpData1(
...
+  const auto &detectorInfo = inpWkSpace->detectorInfo();
   for (int i = 2; i < m_NeighborIDs[1]; i++) {
     int DetID = m_NeighborIDs[i];
 
     size_t workspaceIndex;
-    if (m_wi_to_detid_map.count(DetID) > 0)
-      workspaceIndex = m_wi_to_detid_map.find(DetID)->second;
-    else {
+    try {
+      workspaceIndex = detectorInfo.indexOf(DetID);
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/IntegratePeakTimeSlices.cpp]: See above.

File: Framework/Crystal/test/IntegratePeakTimeSlicesTest.h [hunk 5]
```diff
@@ -97,44 +99,36 @@ class IntegratePeakTimeSlicesTest : public CxxTest::TestSuite {
...
-    double dummy, phi;
-    pos1.getSpherical(dummy, phi, dummy);
-    double ScatAng = phi / 180 * M_PI;
+    const auto &spectrumInfo = wsPtr->spectrumInfo();
+    const auto &detectorInfo = wsPtr->detectorInfo();
+    const auto detID = pixelp->getID();
+    const auto wsIndex = detectorInfo.indexOf(detID);
+    const double L1 = spectrumInfo.l1();
+    double L2 = spectrumInfo.l2(wsIndex);
+    double ScatAng = detectorInfo.twoTheta(wsIndex) / 180 * M_PI;
```
Comment (SimonHeybrock) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: See above. `DetectorInfo::indexOf` does *not* return a workspace index. Doesn't `SpectrumInfo` simply give you what you need here?

File: Framework/Crystal/test/IntegratePeakTimeSlicesTest.h [hunk 6]
```diff
@@ -97,44 +99,36 @@ class IntegratePeakTimeSlicesTest : public CxxTest::TestSuite {
...
-    const detid2index_map map = wsPtr->getDetectorIDToWorkspaceIndexMap(true);
-
     for (int row = 0; row < NRC; row++)
       for (int col = 0; col < NRC; col++) {
 
         boost::shared_ptr<Detector> detP = bankR->getAtXY(col, row);
-
-        detid2index_map::const_iterator it = map.find(detP->getID());
-        size_t wsIndex = (*it).second;
+        const auto wsIndex = detectorInfo.indexOf(detP->getID());
```
Comment (SimonHeybrock) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: See above.

File: Framework/Crystal/test/IntegratePeakTimeSlicesTest.h [hunk 7]
```diff
@@ -237,18 +233,16 @@ class IntegratePeakTimeSlicesTest : public CxxTest::TestSuite {
...
    *   Calculates Q
    */
-  double calcQ(RectangularDetector_const_sptr bankP,
-               boost::shared_ptr<const Instrument> instPtr, int row, int col,
+  double calcQ(const DetectorInfo &detectorInfo, const detid_t detID,
                double time) {
-    boost::shared_ptr<Detector> detP = bankP->getAtXY(col, row);
-
-    double L2 = detP->getDistance(*(instPtr->getSample()));
+    const auto wsIndex = detectorInfo.indexOf(detID);
```
Comment (SimonHeybrock) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: This is not the workspace index, rename the variable.

File: Framework/Crystal/test/IntegratePeakTimeSlicesTest.h [hunk 8]
```diff
@@ -237,18 +233,16 @@ class IntegratePeakTimeSlicesTest : public CxxTest::TestSuite {
...
+    const auto L2 = detectorInfo.l2(wsIndex);
 
     Kernel::Units::MomentumTransfer Q;
     std::vector<double> x;
     x.push_back(time);
-    double L1 = instPtr->getSample()->getDistance(*(instPtr->getSource()));
-    Kernel::V3D pos = detP->getPos();
+    const auto L1 = detectorInfo.l1();
+    const auto pos = detectorInfo.position(wsIndex);
     double ScatAng = fabs(asin(pos.Z() / pos.norm()));
```
Comment (SimonHeybrock) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: Probably you can use `DetectorInfo::twoTheta` to get this simpler?

**Review by SimonHeybrock — COMMENTED** [review]
Some comments below:

File: Framework/Crystal/src/IntegratePeakTimeSlices.cpp (hunk 3 above)
Comment (soininen) [inline Framework/Crystal/src/IntegratePeakTimeSlices.cpp]: Will do.

File: Framework/Crystal/inc/MantidCrystal/FindSXPeaks.h (hunk 2 above)
Comment (SimonHeybrock) [inline Framework/Crystal/inc/MantidCrystal/FindSXPeaks.h]: The test you added forbids detector groups (spectra with more than 1 detector), so unless that is the behavior of the original code you cannot add this test. Maybe a check for `spectrumInfo.hasDetectors(i)` would make sense.

File: Framework/Crystal/test/IntegratePeakTimeSlicesTest.h (hunk 5 above)
Comment (soininen) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: L1 and sample position, yes. I think I should leave the rest as-is as the L2 has to be calculated for an individual detector and they anyway need the phi angle which is not supplied by SpectrumInfo.

File: Framework/Crystal/test/IntegratePeakTimeSlicesTest.h (hunk 8 above)
Comment (soininen) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: Not easily. They refer to the detector by its row and column in a detector bank. I could pass DetectorInfo to the method and get the L1 from it, though, but is it really worth it?

File: Framework/Crystal/test/IntegratePeakTimeSlicesTest.h (hunk 5 above)
Comment (soininen) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: I confused the `index` in `DetectorInfo` with workspace indices as they have the same name and `size_t` type. Hopefully I got it more correct now.

File: Framework/Crystal/inc/MantidCrystal/FindSXPeaks.h (hunk 2 above)
Comment (soininen) [inline Framework/Crystal/inc/MantidCrystal/FindSXPeaks.h]: Good poin. I'll change the test to `hasDetectors()`, then.

File: Framework/Crystal/inc/MantidCrystal/FindSXPeaks.h [hunk 9]
```diff
@@ -69,17 +84,6 @@ struct DLLExport SXPeak {
   @return q vector
   */
   Mantid::Kernel::V3D getQ() const {
-    auto samplePos = _inst->getSample()->getPos();
-    auto sourcePos = _inst->getSource()->getPos();
-    auto detPos = _inst->getDetector(_detectorId)->getPos();
```
Comment (SimonHeybrock) [inline Framework/Crystal/inc/MantidCrystal/FindSXPeaks.h]: Looking at this gain, I think this class should actually use `DetectorInfo`, not `SpectrumInfo`, if we want to keep it consistent with the old behavior (but read also below first!). The index you pass would then be an index of the detector (obtained with `DetectorInfo::indexOf(detid_t)`).

File: Framework/Crystal/src/FindSXPeaks.cpp [hunk 10]
```diff
@@ -155,34 +152,20 @@ void FindSXPeaks::exec() {
...
 
-    double th2 = det->getTwoTheta(sample, L1);
-
     std::vector<int> specs(1, i);
 
-    Mantid::Kernel::V3D L2 = det->getPos();
-    L2 -= sample;
-
-    SXPeak peak(tof, th2, phi, *maxY, specs, l1 + L2.norm(), det->getID(),
-                localworkspace->getInstrument());
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/FindSXPeaks.cpp]: This is a weird: If the spectrum in the workspace has grouped detectors, `localworkspace->getDetector(static_cast<size_t>(i));` above returns a `DetectorGroup`. `det->getID()` will then return the ID of the first detector in the group, so `SXPeak` would use the position of only that first detector! All the other parameters that are passed to the constructor here, such as `th2` are for the whole group!

So in conclusion, I think the way you changed `SXPeak` (to use `SpectrumInfo`) makes sense, since actually there is a bug here (please double check yourself, maybe I am overlooking something!)? Maybe you can open an issue for this with a short description and mark it as fixed by this PR?

**Review by SimonHeybrock — COMMENTED** [review]
I think this looks good now, but read below first. If you can confirm, then this is ready:

File: Framework/Crystal/src/FindSXPeaks.cpp (hunk 10 above)
Comment (soininen) [inline Framework/Crystal/src/FindSXPeaks.cpp]: You are correct with the detector group vs. single detector stuff here. For me it seems this is a problem with `Mantid::Geometry::IPeak` which is not designed to work with detector groups. The algorithm outputs a `PeaksWorkspace`, which consists of `IPeak`s which in turn hold information like *Q*, *Intensity* but also detector ID. Now, with my changes, the *Q* and *Intensity* can be for a detector group but the ID is still for a single detector. This could be fixed by getting these values for single detectors, but I suppose there is no way of getting *Intensity* for a single detector in a group, is there? We could also check for `hasUniqueDetector()` here and give warning/bail out but would this change in behavior be undesirable?

With this single detector ID going to the output workspace, I don't think I actually fixed anything here. I could switch to `DetectorInfo`, but then again the intensity data would come from the spectrum itself, if I understood things right. Maybe it would be the right thing to keep as close to the old behavior as possible? Let me know your thoughts about this.

Comment (SimonHeybrock) [inline Framework/Crystal/src/FindSXPeaks.cpp]: Yes, good point, the intensity is only available for the group. Maybe try adding the check for `hasUniqueDetector()` that you suggested and see if any test fail?

@mantidproject/gatekeepers Could someone else please have a look at this problem, I don't think I have enough knowledge of what is going on with peaks in Mantid to make a final decision here.

Comment (soininen) [inline Framework/Crystal/src/FindSXPeaks.cpp]: All tests pass even if the spectra are tested for unique detectors.

Comment (martyngigg) [inline Framework/Crystal/src/FindSXPeaks.cpp]: I'm not most expert on the peaks side of things so some one else should comment for that but in the case of a group of detectors it does not make any sense to try and pull out an intensity from an individual detector - that information has been lost when the grouping is performed (whether hardware or software). 

@soininen Do the tests actually use any grouping anywhere?

Comment (soininen) [inline Framework/Crystal/src/FindSXPeaks.cpp]: @martyngigg By a quick look, the tests don't seem to be using detector groups.

**Comment by SimonHeybrock** [comment]
@soininen Since no one else is commenting regarding `FindSXPeaks`, I would suggest to add the test for unique detectors, and then I'll put the shipit.

**Comment by SimonHeybrock** [comment]
Looks good, :shipit: once builds green.
